### PR TITLE
Quilt Generic PR Policy

### PR DESCRIPTION
**Pull request by TheGlitch76**
Third time's the charm, hopefully!
This one is generic to projects, and I intend to apply it to both Loader and QSL immediately.

Since we need to define a policy for QSL quickly, because PRs are beginning to ramp up, let's try to get this reviewed and accepted in a timely manner.

[Rendered View](https://github.com/theglitch76-forks/rfcs/blob/generic-pr-policy/structure/0039-pr-policy.md)

### CODE REVIEW

**Review by LambdAurora — APPROVED** [review]

**Comment by BasiqueEvangelist** [comment]
Can you add a link to the rendered view?

**Review by grondag — APPROVED** [review]

File: structure/0039-pr-policy.md
Comment (LeoCTH) [inline structure/0039-pr-policy.md:30]: critera -> criteria

**Review by LeoCTH — APPROVED** [review]

**Review by Leo40Git — APPROVED** [review]

File: structure/0039-pr-policy.md
Comment (OroArmor) [inline structure/0039-pr-policy.md:27]: This is a little vauge. Do you mean to say that no team members have a pending review?

**Review by OroArmor — COMMENTED** [review]
I have some questions about the Template section:
For QM would this work:

----

To get a pull request merged into ___Quilt Mappings___, it must get ***a minimum of 2*** approvals from ***Mappings Triage*** members ***and at least one from a Mappings Team member,*** and then enter a Final Comment Period to ensure everyone has had a chance to look at the PR. Once these criteria have been met, the PR will be merged. 

#### New Mappings
**Required Approvals**: ***2 From Mappings Triage, at least one from a Mappings Team member***

**Final Comment Period**: ***Merged before/when the next snapshot is released***

----
all my additions are in bold and italics
... include sections for backports and refactors

File: structure/0039-pr-policy.md
Comment (TheGlitch76) [reply structure/0039-pr-policy.md:27]: I'm referring to the "request review" feature on Github.

**Comment by TheGlitch76** [comment]
@OroArmor Yes, that's exactly what I'm looking for. It's fine to be flexible with your specific criteria, this RFC is a guideline and not a strict template.

**Comment by TheGlitch76** [comment]
I'm requesting an admin to review this and potentially put it on FCP as there's no team that specifically has responsibility for this RFC.

File: structure/0039-pr-policy.md
Comment (i509VCB) [inline structure/0039-pr-policy.md:18]: An approval should be defined as using the review feature of github to make it more clear who is approving.

**Review by i509VCB — APPROVED** [review]
Other than the definition of a "approving review" being defined to make this public to github, everything else is fine.

File: structure/0039-pr-policy.md
Comment (ByMartrixx) [inline structure/0039-pr-policy.md:42]: 't' in `PROJECt` should be uppercase

**Review by ByMartrixx — APPROVED** [review]
Aside from my comment, everything looks good to me

**Review by OroArmor — APPROVED** [review]

**Review by LemmaEOF — APPROVED** [review]

**Comment by LemmaEOF** [comment]
LGTM/LGTM

**Comment by Haven-King** [comment]
Moving this to FCP, set to conclude on November 21st at 1:00AM GMT.

File: structure/0039-pr-policy.md
Comment (sciwhiz12) [inline structure/0039-pr-policy.md:30]: Let us take a hypothetical scenario where a PR has been approved through this policy and has been merged, and then a critical issue is found in the implementation of the PR which needs immediate correcting. Let us then say that the same PR author creates a new PR immediately after the discovery of said bug in order to fix or remedy it in some way. 

Let us take that this RFC is passed as it currently stands:
 - If the PR is subjected to the same process as defined in this RFC as augmented by the repository's contributing information, then the fix for said hypothetical critical issue would be delayed for at least the time for the final comment period.
 - If a maintainer of the repository manually commits the same fix as contributed by the PR, it would essentially be an act of fiat from the maintainer to bypass the PR process and manually integrate the PR's changes, and sets the same precedent within the project.

Therefore, either the fix for the critical issue is delayed, or a precedent of maintainers being able to effectively bypass this PR policy is set outside of this RFC.

As a solution for this hypothetical scenario, I suggest a clause be added wherein if the active members of the team show unanimous consensus for the merging of a PR, with some additional condition(s) to prevent the abuse of this by the team (such as agreement by a technical lead and/or Quilt admin), then the PR may be merged bypassing the final comment period.

File: structure/0039-pr-policy.md
Comment (sciwhiz12) [inline structure/0039-pr-policy.md:38]: Would it be acceptable to add a clause to this which requires the dismisser of a review to add their reason for the review's dismissal (such as the last known time of contact for that reviewer) in the reason box provided by GitHub when dismissing a review?

File: structure/0039-pr-policy.md
Comment (TheGlitch76) [reply structure/0039-pr-policy.md:30]: I don't think we need to have a specific defined method of going around this RFC, but it should be left to the individual teams on how they wish to approach this problem.

For example, in my proposed PR policy for QSL, I add an "urgent" triage category that has no FCP for this exact use case. Additionally, I define in the policy that those with push access may bypass the policy for trivial changes like typos after double checking with the team that they think it's okay. Technically, two bad actors could merge whatever they want by abusing the `urgent` tag, (or just solo pushing whatever they want) but it's unlikely that those people would remain members of the team for very long.

Other projects may want to make different decisions about these extraneous circumstances, and that's fine! This process is more of a guideline than a strict definition.

File: structure/0039-pr-policy.md
Comment (TheGlitch76) [reply structure/0039-pr-policy.md:31]: Good catch--it will begin immediately when we soon:tm: have a bot that automates this process, but it shouldn't start silently.

File: structure/0039-pr-policy.md
Comment (TheGlitch76) [reply structure/0039-pr-policy.md:60]: > Also, who can request such a vote to be held

This is a good question to ask. Any team member responsible for the PR may request that vote. I'll clarify it in the RFC.

>  Also, who can request such a vote to be held? Is there a minimum period of time after PR creation where such a vote cannot be requested (to prevent those which can request a vote from pushing it forward without due consideration, in the hopes that it will be rejected by the vote)

No, and I don't think that this RFC should focus on trying to avoid people acting in bad faith. If someone is abusing this process to push an agenda, they should be removed from the team.
However, I do see that it would be useful to be able to vote that the PR still needs time to mature. I'll add a clause about that.

File: structure/0039-pr-policy.md
Comment (TheGlitch76) [reply structure/0039-pr-policy.md:38]: Yes, this makes sense.

File: structure/0039-pr-policy.md
Comment (TheGlitch76) [reply structure/0039-pr-policy.md:61]: I though about this a bit further. What I decided on doing is making the voting FCP last two weeks (to give everyone plenty of time to talk out disagreements, make changes to the PR, etc), and adding an option to vote to keep the PR open.

File: structure/0039-pr-policy.md
Comment (sciwhiz12) [inline structure/0039-pr-policy.md:43]: The proper language identifier here for Markdown is `markdown`, not `md` (it seems to be for some sort of LISP-related language from my small investigation).

**Review by sciwhiz12 — APPROVED** [review]
Looks good to me now, with only a small quibble regarding the code block.

File: structure/0039-pr-policy.md
Comment (williambl) [inline structure/0039-pr-policy.md:27]: IMO this should be reworded to specify that there are no requested reviews _still pending_ - as it is, it reads (at least to me) as if requesting a review makes your PR invalid.

Comment (williambl) [reply structure/0039-pr-policy.md:27]: " - Everybody in the team who has had their review requested on the Pull Request using the GitHub request review system has submitted a review" or something, perhaps?

**Pull request merged**
